{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "61721487-dd83-45fb-b940-4de5235c3214",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from tdc.multi_pred import DTI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "58ee4cbe-89c3-425e-8d50-4111ca029ba3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "My numpy version is:  1.26.2\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "print(\"My numpy version is: \", np.__version__)\n",
    "print(\"My numpy version is: \", np.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5dafa10-482a-42d3-9e8e-67465f1bf3a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.9.12\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "major = sys.version_info.major\n",
    "minor = sys.version_info.minor\n",
    "micro = sys.version_info.micro\n",
    "\n",
    "print(f\"Python version: {major}.{minor}.{micro}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "58479a37-245d-4173-9580-ff319b641a89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pandas in /opt/anaconda3/lib/python3.11/site-packages (2.2.3)\n",
      "Requirement already satisfied: numpy in /opt/anaconda3/lib/python3.11/site-packages (1.26.4)\n",
      "Collecting numpy\n",
      "  Using cached numpy-2.1.2-cp311-cp311-macosx_14_0_arm64.whl.metadata (60 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /opt/anaconda3/lib/python3.11/site-packages (from pandas) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/lib/python3.11/site-packages (from pandas) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/lib/python3.11/site-packages (from pandas) (2023.3)\n",
      "Requirement already satisfied: six>=1.5 in /opt/anaconda3/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas) (1.16.0)\n",
      "Using cached numpy-2.1.2-cp311-cp311-macosx_14_0_arm64.whl (5.4 MB)\n",
      "Installing collected packages: numpy\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.26.4\n",
      "    Uninstalling numpy-1.26.4:\n",
      "      Successfully uninstalled numpy-1.26.4\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "pytdc 0.3.8 requires rdkit-pypi, which is not installed.\n",
      "gensim 4.3.0 requires FuzzyTM>=0.4.0, which is not installed.\n",
      "streamlit 1.30.0 requires numpy<2,>=1.19.3, but you have numpy 2.1.2 which is incompatible.\n",
      "contourpy 1.2.0 requires numpy<2.0,>=1.20, but you have numpy 2.1.2 which is incompatible.\n",
      "matplotlib 3.8.0 requires numpy<2,>=1.21, but you have numpy 2.1.2 which is incompatible.\n",
      "astropy 5.3.4 requires numpy<2,>=1.21, but you have numpy 2.1.2 which is incompatible.\n",
      "numba 0.59.0 requires numpy<1.27,>=1.22, but you have numpy 2.1.2 which is incompatible.\n",
      "pywavelets 1.5.0 requires numpy<2.0,>=1.22.4, but you have numpy 2.1.2 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed numpy-2.1.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Loading...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade pandas numpy\n",
    "data = DTI(name='KIBA')\n",
    "# KibaDataSet = data.get_data()\n",
    "# KibaDataSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e5b5cacb-ea55-4b11-95f6-8abe24198add",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug_ID</th>\n",
       "      <th>Drug</th>\n",
       "      <th>Target_ID</th>\n",
       "      <th>Target</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL1087421</td>\n",
       "      <td>COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2</td>\n",
       "      <td>O00141</td>\n",
       "      <td>MTVKTEAAKGTLTYSRMRGMVAILIAFMKQRRMGLNDFIQKIANNS...</td>\n",
       "      <td>11.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL1087421</td>\n",
       "      <td>COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2</td>\n",
       "      <td>O14920</td>\n",
       "      <td>MSWSPSLTTQTCGAWEMKERLGTGGFGNVIRWHNQETGEQIAIKQC...</td>\n",
       "      <td>11.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1087421</td>\n",
       "      <td>COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2</td>\n",
       "      <td>O15111</td>\n",
       "      <td>MERPPGLRPGAGGPWEMRERLGTGGFGNVCLYQHRELDLKIAIKSC...</td>\n",
       "      <td>11.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1087421</td>\n",
       "      <td>COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2</td>\n",
       "      <td>P00533</td>\n",
       "      <td>MRPSGTAGAALLALLAALCPASRALEEKKVCQGTSNKLTQLGTFED...</td>\n",
       "      <td>11.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1087421</td>\n",
       "      <td>COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2</td>\n",
       "      <td>P04626</td>\n",
       "      <td>MELAALCRWGLLLALLPPGAASTQVCTGTDMKLRLPASPETHLDML...</td>\n",
       "      <td>11.10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117652</th>\n",
       "      <td>CHEMBL230654</td>\n",
       "      <td>CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...</td>\n",
       "      <td>Q13554</td>\n",
       "      <td>MATTVTCTRFTDEYQLYEDIGKGAFSVVRRCVKLCTGHEYAAKIIN...</td>\n",
       "      <td>10.49794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117653</th>\n",
       "      <td>CHEMBL230654</td>\n",
       "      <td>CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...</td>\n",
       "      <td>Q13555</td>\n",
       "      <td>MATTATCTRFTDDYQLFEELGKGAFSVVRRCVKKTSTQEYAAKIIN...</td>\n",
       "      <td>10.49794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117654</th>\n",
       "      <td>CHEMBL230654</td>\n",
       "      <td>CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...</td>\n",
       "      <td>Q13557</td>\n",
       "      <td>MASTTTCTRFTDEYQLFEELGKGAFSVVRRCMKIPTGQEYAAKIIN...</td>\n",
       "      <td>10.49794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117655</th>\n",
       "      <td>CHEMBL230654</td>\n",
       "      <td>CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...</td>\n",
       "      <td>Q16539</td>\n",
       "      <td>MSQERPTFYRQELNKTIWEVPERYQNLSPVGSGAYGSVCAAFDTKT...</td>\n",
       "      <td>10.49794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117656</th>\n",
       "      <td>CHEMBL230654</td>\n",
       "      <td>CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...</td>\n",
       "      <td>Q9UQM7</td>\n",
       "      <td>MATITCTRFTEEYQLFEELGKGAFSVVRRCVKVLAGQEYAAKIINT...</td>\n",
       "      <td>10.49794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117657 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Drug_ID                                               Drug  \\\n",
       "0       CHEMBL1087421            COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2   \n",
       "1       CHEMBL1087421            COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2   \n",
       "2       CHEMBL1087421            COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2   \n",
       "3       CHEMBL1087421            COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2   \n",
       "4       CHEMBL1087421            COc1cc2c(cc1Cl)C(c1ccc(Cl)c(Cl)c1)=NCC2   \n",
       "...               ...                                                ...   \n",
       "117652   CHEMBL230654  CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...   \n",
       "117653   CHEMBL230654  CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...   \n",
       "117654   CHEMBL230654  CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...   \n",
       "117655   CHEMBL230654  CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...   \n",
       "117656   CHEMBL230654  CCCc1nc[nH]c1CNc1cc(Cl)c2ncc(C#N)c(Nc3ccc(F)c(...   \n",
       "\n",
       "       Target_ID                                             Target         Y  \n",
       "0         O00141  MTVKTEAAKGTLTYSRMRGMVAILIAFMKQRRMGLNDFIQKIANNS...  11.10000  \n",
       "1         O14920  MSWSPSLTTQTCGAWEMKERLGTGGFGNVIRWHNQETGEQIAIKQC...  11.10000  \n",
       "2         O15111  MERPPGLRPGAGGPWEMRERLGTGGFGNVCLYQHRELDLKIAIKSC...  11.10000  \n",
       "3         P00533  MRPSGTAGAALLALLAALCPASRALEEKKVCQGTSNKLTQLGTFED...  11.10000  \n",
       "4         P04626  MELAALCRWGLLLALLPPGAASTQVCTGTDMKLRLPASPETHLDML...  11.10000  \n",
       "...          ...                                                ...       ...  \n",
       "117652    Q13554  MATTVTCTRFTDEYQLYEDIGKGAFSVVRRCVKLCTGHEYAAKIIN...  10.49794  \n",
       "117653    Q13555  MATTATCTRFTDDYQLFEELGKGAFSVVRRCVKKTSTQEYAAKIIN...  10.49794  \n",
       "117654    Q13557  MASTTTCTRFTDEYQLFEELGKGAFSVVRRCMKIPTGQEYAAKIIN...  10.49794  \n",
       "117655    Q16539  MSQERPTFYRQELNKTIWEVPERYQNLSPVGSGAYGSVCAAFDTKT...  10.49794  \n",
       "117656    Q9UQM7  MATITCTRFTEEYQLFEELGKGAFSVVRRCVKVLAGQEYAAKIINT...  10.49794  \n",
       "\n",
       "[117657 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KibaDataSet = data.get_data()\n",
    "KibaDataSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "64c6fddc-f532-4405-a8e7-eee03e57c5b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "229"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(KibaDataSet[\"Target_ID\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc14169-5537-4d56-88c8-a357d3deabb9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "388da081-edb0-45be-84ee-71a2ef5a3e6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found local copy...\n",
      "Loading...\n",
      "Done!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Valid sequences have been saved to individual text files in ./KibaValidSequences\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "\n",
    "def extract_sequences_from_kiba(kiba_data):\n",
    "    \"\"\"\n",
    "    Extracts protein sequences from the KIBA dataset.\n",
    "    \n",
    "    Parameters:\n",
    "    kiba_data (pd.DataFrame): The KIBA dataset.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are Target_IDs and values are the protein sequences.\n",
    "    \"\"\"\n",
    "    sequence_dict = {}\n",
    "    for idx, row in kiba_data.iterrows():\n",
    "        target_id = row['Target_ID']\n",
    "        sequence = row['Target']\n",
    "        sequence_dict[target_id] = sequence\n",
    "    return sequence_dict\n",
    "\n",
    "def verify_sequence_length_with_contact_map(sequence_dict, contact_map_folder):\n",
    "    \"\"\"\n",
    "    Verifies that the sequence lengths from KIBA match the dimensions of the contact maps.\n",
    "    \n",
    "    Parameters:\n",
    "    sequence_dict (dict): Dictionary containing KIBA protein sequences.\n",
    "    contact_map_folder (str): Path to the folder containing the contact maps (.npy files).\n",
    "    \n",
    "    Returns:\n",
    "    dict: A dictionary of valid sequences where the sequence length matches the contact map dimensions.\n",
    "    \"\"\"\n",
    "    valid_sequence_dict = {}\n",
    "    mismatches = []\n",
    "\n",
    "    for contact_file in os.listdir(contact_map_folder):\n",
    "        if contact_file.endswith('.npy'):\n",
    "            protein_id = contact_file.split('.npy')[0]\n",
    "            contact_map_path = os.path.join(contact_map_folder, contact_file)\n",
    "            contact_map = np.load(contact_map_path)\n",
    "\n",
    "            # Check if we have a sequence for this protein\n",
    "            if protein_id in sequence_dict:\n",
    "                sequence_length = len(sequence_dict[protein_id])\n",
    "                contact_map_shape = contact_map.shape[0]\n",
    "\n",
    "                # Ensure sequence length matches contact map dimensions\n",
    "                if sequence_length == contact_map_shape:\n",
    "                    valid_sequence_dict[protein_id] = sequence_dict[protein_id]\n",
    "                else:\n",
    "                    mismatches.append(f\"Mismatch for {protein_id}: Sequence length = {sequence_length}, Contact map shape = {contact_map_shape}\")\n",
    "            else:\n",
    "                mismatches.append(f\"Target ID {protein_id} not found in KIBA dataset\")\n",
    "    \n",
    "    return valid_sequence_dict, mismatches\n",
    "\n",
    "# Example usage\n",
    "from tdc.multi_pred import DTI\n",
    "data = DTI(name='KIBA')\n",
    "kiba_dataset = data.get_data()\n",
    "\n",
    "# Step 1: Extract sequences from the KIBA dataset\n",
    "sequence_dict = extract_sequences_from_kiba(kiba_dataset)\n",
    "\n",
    "# Step 2: Verify that KIBA sequence lengths match the contact map dimensions\n",
    "contact_map_folder = \"./KibaContactMaps\"\n",
    "valid_sequence_dict, mismatches = verify_sequence_length_with_contact_map(sequence_dict, contact_map_folder)\n",
    "\n",
    "# Output mismatches\n",
    "if mismatches:\n",
    "    print(\"Mismatches found:\")\n",
    "    for mismatch in mismatches:\n",
    "        print(mismatch)\n",
    "\n",
    "# Save the valid sequences to individual text files (if needed)\n",
    "output_folder = \"./KibaValidSequences\"\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)\n",
    "\n",
    "for protein_id, sequence in valid_sequence_dict.items():\n",
    "    file_path = os.path.join(output_folder, f\"{protein_id}.txt\")\n",
    "    with open(file_path, 'w') as f:\n",
    "        f.write(sequence)\n",
    "\n",
    "print(f\"Valid sequences have been saved to individual text files in {output_folder}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7eeb9aa-0234-40f2-8990-0568f74f2263",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6974fe43-b0e8-461f-a146-9ab522df39c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# valid_sequence_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "17bc9c50-9a52-47e4-911d-6c2a0f0ec834",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import os\n",
    "# import numpy as np\n",
    "# import torch\n",
    "# import torch_geometric.data as DATA\n",
    "# from torch_geometric.utils import from_networkx\n",
    "# import networkx as nx\n",
    "\n",
    "# # Residue properties for feature generation\n",
    "\n",
    "# res_weight_table = {'A': 71.08, 'C': 103.15, 'D': 115.09, 'E': 129.12, 'F': 147.18, 'G': 57.05, 'H': 137.14,\n",
    "#                     'I': 113.16, 'K': 128.18, 'L': 113.16, 'M': 131.20, 'N': 114.11, 'P': 97.12, 'Q': 128.13,\n",
    "#                     'R': 156.19, 'S': 87.08, 'T': 101.11, 'V': 99.13, 'W': 186.22, 'Y': 163.18}\n",
    "\n",
    "# res_pka_table = {'A': 2.34, 'C': 1.96, 'D': 1.88, 'E': 2.19, 'F': 1.83, 'G': 2.34, 'H': 1.82, 'I': 2.36,\n",
    "#                  'K': 2.18, 'L': 2.36, 'M': 2.28, 'N': 2.02, 'P': 1.99, 'Q': 2.17, 'R': 2.17, 'S': 2.21,\n",
    "#                  'T': 2.09, 'V': 2.32, 'W': 2.83, 'Y': 2.32}\n",
    "\n",
    "# res_pkb_table = {'A': 9.69, 'C': 10.28, 'D': 9.60, 'E': 9.67, 'F': 9.13, 'G': 9.60, 'H': 9.17,\n",
    "#                  'I': 9.60, 'K': 8.95, 'L': 9.60, 'M': 9.21, 'N': 8.80, 'P': 10.60, 'Q': 9.13,\n",
    "#                  'R': 9.04, 'S': 9.15, 'T': 9.10, 'V': 9.62, 'W': 9.39, 'Y': 9.62}\n",
    "\n",
    "# res_pkx_table = {'A': 0.00, 'C': 8.18, 'D': 3.65, 'E': 4.25, 'F': 0.00, 'G': 0, 'H': 6.00,\n",
    "#                  'I': 0.00, 'K': 10.53, 'L': 0.00, 'M': 0.00, 'N': 0.00, 'P': 0.00, 'Q': 0.00,\n",
    "#                  'R': 12.48, 'S': 0.00, 'T': 0.00, 'V': 0.00, 'W': 0.00, 'Y': 0.00}\n",
    "\n",
    "# res_pl_table = {'A': 6.00, 'C': 5.07, 'D': 2.77, 'E': 3.22, 'F': 5.48, 'G': 5.97, 'H': 7.59,\n",
    "#                 'I': 6.02, 'K': 9.74, 'L': 5.98, 'M': 5.74, 'N': 5.41, 'P': 6.30, 'Q': 5.65,\n",
    "#                 'R': 10.76, 'S': 5.68, 'T': 5.60, 'V': 5.96, 'W': 5.89, 'Y': 5.96}\n",
    "\n",
    "# res_hydrophobic_ph2_table = {'A': 47, 'C': 52, 'D': -18, 'E': 8, 'F': 92, 'G': 0, 'H': -42, 'I': 100,\n",
    "#                              'K': -37, 'L': 100, 'M': 74, 'N': -41, 'P': -46, 'Q': -18, 'R': -26, 'S': -7,\n",
    "#                              'T': 13, 'V': 79, 'W': 84, 'Y': 49}\n",
    "\n",
    "# res_hydrophobic_ph7_table = {'A': 41, 'C': 49, 'D': -55, 'E': -31, 'F': 100, 'G': 0, 'H': 8, 'I': 99,\n",
    "#                              'K': -23, 'L': 97, 'M': 74, 'N': -28, 'P': -46, 'Q': -10, 'R': -14, 'S': -5,\n",
    "#                              'T': 13, 'V': 76, 'W': 97, 'Y': 63}\n",
    "\n",
    "# # Residue group properties\n",
    "# aliphatic = ['A', 'I', 'L', 'M', 'V']\n",
    "# aromatic = ['F', 'W', 'Y']\n",
    "# polar_neutral = ['C', 'N', 'Q', 'S', 'T']\n",
    "# acidic_charged = ['D', 'E']\n",
    "# basic_charged = ['H', 'K', 'R']\n",
    "\n",
    "# # 1. Function to generate the full node feature vector for each residue\n",
    "\n",
    "# def residue_features(residue):\n",
    "#     \"\"\"\n",
    "#     Generates a feature vector for each residue (amino acid).\n",
    "    \n",
    "#     Parameters:\n",
    "#     residue (str): The single-letter amino acid code (e.g., 'A', 'C').\n",
    "\n",
    "#     Returns:\n",
    "#     numpy array: A 54-dimensional feature vector for the residue.\n",
    "#     \"\"\"\n",
    "#     # One-hot encoding of the residue (21-dimensional)\n",
    "#     one_hot = [1 if residue == r else 0 for r in res_weight_table.keys()]\n",
    "\n",
    "#     # Residue group properties (binary flags)\n",
    "#     aliphatic_flag = [1] if residue in aliphatic else [0]\n",
    "#     aromatic_flag = [1] if residue in aromatic else [0]\n",
    "#     polar_neutral_flag = [1] if residue in polar_neutral else [0]\n",
    "#     acidic_charged_flag = [1] if residue in acidic_charged else [0]\n",
    "#     basic_charged_flag = [1] if residue in basic_charged else [0]\n",
    "\n",
    "#     # Chemical properties (single scalar values)\n",
    "#     weight = [res_weight_table.get(residue, 0)]  # Molecular weight\n",
    "#     pka = [res_pka_table.get(residue, 0)]  # pKa of -COOH\n",
    "#     pkb = [res_pkb_table.get(residue, 0)]  # pKb of -NH3\n",
    "#     pkx = [res_pkx_table.get(residue, 0)]  # pKa of side chain\n",
    "#     pi = [res_pl_table.get(residue, 0)]  # Isoelectric point (pI)\n",
    "#     hydrophobicity_ph2 = [res_hydrophobic_ph2_table.get(residue, 0)]  # Hydrophobicity at pH 2\n",
    "#     hydrophobicity_ph7 = [res_hydrophobic_ph7_table.get(residue, 0)]  # Hydrophobicity at pH 7\n",
    "\n",
    "#     # Combine all features into a single vector (54-dimensional)\n",
    "#     feature_vector = (\n",
    "#         one_hot +\n",
    "#         aliphatic_flag +\n",
    "#         aromatic_flag +\n",
    "#         polar_neutral_flag +\n",
    "#         acidic_charged_flag +\n",
    "#         basic_charged_flag +\n",
    "#         weight +\n",
    "#         pka +\n",
    "#         pkb +\n",
    "#         pkx +\n",
    "#         pi +\n",
    "#         hydrophobicity_ph2 +\n",
    "#         hydrophobicity_ph7\n",
    "#     )\n",
    "    \n",
    "#     return np.array(feature_vector)\n",
    "\n",
    "# # 2. Function to create the graph from the contact map and node features\n",
    "\n",
    "# def create_graph_from_contact_map(protein_sequence, contact_map):\n",
    "#     \"\"\"\n",
    "#     Creates a graph from a protein contact map, adding node features based on residue properties.\n",
    "\n",
    "#     Parameters:\n",
    "#     protein_sequence (str): The amino acid sequence of the protein.\n",
    "#     contact_map (numpy array): The NxN contact map matrix where each entry indicates a contact.\n",
    "\n",
    "#     Returns:\n",
    "#     torch_geometric.data.Data: The graph data in PyTorch Geometric format.\n",
    "#     \"\"\"\n",
    "#     num_residues = len(protein_sequence)  # Number of residues in the sequence\n",
    "\n",
    "#     # Create a NetworkX graph\n",
    "#     G = nx.Graph()\n",
    "\n",
    "#     # Add nodes with residue features\n",
    "#     for i, residue in enumerate(protein_sequence):\n",
    "#         G.add_node(i, features=residue_features(residue))\n",
    "\n",
    "#     # Add edges based on the contact map (threshold: contact_map >= 0.5)\n",
    "#     for i in range(num_residues):\n",
    "#         for j in range(i + 1, num_residues):\n",
    "#             if contact_map[i, j] >= 0.5:  # Threshold for contact :todo: chnage it nd tune it for better results !!\n",
    "#                 G.add_edge(i, j)\n",
    "\n",
    "#     # Convert the NetworkX graph to PyTorch Geometric format\n",
    "#     data = from_networkx(G)\n",
    "\n",
    "#     # Extract node features into the `x` attribute (used by PyTorch Geometric)\n",
    "#     data.x = torch.tensor([G.nodes[i]['features'] for i in G.nodes], dtype=torch.float)\n",
    "\n",
    "#     return data\n",
    "\n",
    "# # 3. Process the valid sequences and corresponding contact maps\n",
    "\n",
    "# def process_proteins(sequence_dict, contact_map_folder):\n",
    "#     \"\"\"\n",
    "#     Processes all proteins: loads contact maps and creates graphs with node features for each protein.\n",
    "\n",
    "#     Parameters:\n",
    "#     sequence_dict (dict): Dictionary containing the valid protein sequences.\n",
    "#     contact_map_folder (str): Path to the folder containing the contact maps.\n",
    "\n",
    "#     Returns:\n",
    "#     dict: A dictionary where keys are protein IDs and values are graph data objects.\n",
    "#     \"\"\"\n",
    "#     graph_data_dict = {}\n",
    "\n",
    "#     for contact_file in os.listdir(contact_map_folder):\n",
    "#         if contact_file.endswith('.npy'):\n",
    "#             protein_id = contact_file.split('.npy')[0]\n",
    "            \n",
    "#             if protein_id in sequence_dict:\n",
    "#                 # Load the contact map\n",
    "#                 contact_map_path = os.path.join(contact_map_folder, contact_file)\n",
    "#                 contact_map = np.load(contact_map_path)\n",
    "\n",
    "#                 # Retrieve the protein sequence\n",
    "#                 sequence = sequence_dict[protein_id]\n",
    "\n",
    "#                 # Create the graph from the contact map and sequence\n",
    "#                 graph_data = create_graph_from_contact_map(sequence, contact_map)\n",
    "                \n",
    "#                 # Store the graph data in the dictionary\n",
    "#                 graph_data_dict[protein_id] = graph_data\n",
    "#             else:\n",
    "#                 print(f\"Warning: No sequence found for protein {protein_id}\")\n",
    "\n",
    "#     return graph_data_dict\n",
    "\n",
    "# # Example usage\n",
    "\n",
    "# contact_map_folder = \"./KibaContactMaps\"\n",
    "\n",
    "# # Assuming valid_sequence_dict is already populated\n",
    "# graph_data_dict = process_proteins(valid_sequence_dict, contact_map_folder)\n",
    "\n",
    "# # You can now use graph_data_dict for downstream tasks\n",
    "# for protein_id, graph in graph_data_dict.items():\n",
    "#     print(f\"Graph for {protein_id}: {graph}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f8524f-bd99-4435-a36a-98c390e62eed",
   "metadata": {},
   "source": [
    "\\section*{Position-Specific Scoring Matrix (PSSM)}\n",
    "\n",
    "A **Position-Specific Scoring Matrix (PSSM)** is a matrix used in bioinformatics to represent the likelihood of finding a particular amino acid at each position of a protein sequence, based on an alignment of multiple sequences. The matrix is derived from a **multiple sequence alignment (MSA)** of related protein sequences, where each column represents the probabilities of different amino acids appearing at a specific position.\n",
    "\n",
    "Each row in the matrix corresponds to a position in the sequence (i.e., one residue in the protein), and each column corresponds to one of the 20 standard amino acids (with a possible additional column for unknown residues or gaps).\n",
    "\n",
    "\\subsection*{Simple Example}\n",
    "\n",
    "Imagine we have the following **protein sequence**:\n",
    "\n",
    "\\[\n",
    "\\text{Sequence}: A T G C\n",
    "\\]\n",
    "\n",
    "We also have several **aligned sequences** from other similar proteins:\n",
    "\n",
    "\\[\n",
    "\\begin{aligned}\n",
    "&\\text{A T G C} \\\\\n",
    "&\\text{A A G C} \\\\\n",
    "&\\text{T T G C} \\\\\n",
    "&\\text{A T T C} \\\\\n",
    "\\end{aligned}\n",
    "\\]\n",
    "\n",
    "Here, we have four sequences of length 4. For each position in the protein sequence, we count how often each amino acid appears.\n",
    "\n",
    "For example, at the **first position**:\n",
    "- 'A' appears 3 times.\n",
    "- 'T' appears 1 time.\n",
    "\n",
    "At the **second position**:\n",
    "- 'T' appears 3 times.\n",
    "- 'A' appears 1 time.\n",
    "\n",
    "\\subsection*{Position Frequency Matrix (PFM)}\n",
    "\n",
    "The **Position Frequency Matrix (PFM)** is the raw count of amino acid occurrences in each position of the alignment. For the example sequence \\texttt{A T G C}, the PFM looks like:\n",
    "\n",
    "\\[\n",
    "\\begin{array}{c|c|c|c|c}\n",
    "\\text{Position} & A & T & G & C \\\\\n",
    "\\hline\n",
    "1 & 3 & 1 & 0 & 0 \\\\\n",
    "2 & 1 & 3 & 0 & 0 \\\\\n",
    "3 & 0 & 0 & 4 & 0 \\\\\n",
    "4 & 0 & 0 & 0 & 4 \\\\\n",
    "\\end{array}\n",
    "\\]\n",
    "\n",
    "\\subsection*{Adding Pseudocounts}\n",
    "\n",
    "To avoid zero probabilities, we add **pseudocounts** to the matrix. A small value (e.g., 0.8) is added to each count to ensure no entry in the matrix is zero.\n",
    "\n",
    "\\subsection*{Position-Specific Scoring Matrix (PSSM)}\n",
    "\n",
    "To calculate the **PSSM**, we normalize the frequencies by dividing the counts by the total number of sequences (in this case, 4):\n",
    "\n",
    "\\[\n",
    "\\begin{array}{c|c|c|c|c}\n",
    "\\text{Position} & A & T & G & C \\\\\n",
    "\\hline\n",
    "1 & 0.75 & 0.25 & 0.00 & 0.00 \\\\\n",
    "2 & 0.25 & 0.75 & 0.00 & 0.00 \\\\\n",
    "3 & 0.00 & 0.00 & 1.00 & 0.00 \\\\\n",
    "4 & 0.00 & 0.00 & 0.00 & 1.00 \\\\\n",
    "\\end{array}\n",
    "\\]\n",
    "\n",
    "This table shows the likelihood of each amino acid appearing at each position. The PSSM is essentially this matrix of probabilities (though in advanced cases, log-odds values may be used).\n",
    "\n",
    "\\subsection*{How We Calculate PSSM in the Code}\n",
    "\n",
    "In the code, the **PSSM calculation** is performed as follows:\n",
    "\\begin{enumerate}\n",
    "    \\item **Reading an alignment file**: The multiple sequence alignment (MSA) is read from an alignment file.\n",
    "    \\item **Counting occurrences**: For each position in the alignment, the number of occurrences of each amino acid is counted (creating the PFM).\n",
    "    \\item **Adding pseudocounts**: Small pseudocounts are added to avoid zero probabilities.\n",
    "    \\item **Normalizing**: The counts are divided by the total number of sequences to create a probability matrix (PPM), which is treated as the PSSM.\n",
    "\\end{enumerate}\n",
    "\n",
    "\\subsection*{Conclusion}\n",
    "\n",
    "The PSSM matrix provides a way to capture the variability of amino acids at specific positions within an alignment, and it is widely used in various bioinformatics applications such as sequence alignment and protein structure prediction.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1f66c691-b399-4b13-8ee5-a017fae7d738",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created and saved graph 1/203. Remaining: 202\n",
      "Created and saved graph 2/203. Remaining: 201\n",
      "Created and saved graph 3/203. Remaining: 200\n",
      "Created and saved graph 4/203. Remaining: 199\n",
      "Created and saved graph 5/203. Remaining: 198\n",
      "Created and saved graph 6/203. Remaining: 197\n",
      "Created and saved graph 7/203. Remaining: 196\n",
      "Created and saved graph 8/203. Remaining: 195\n",
      "Created and saved graph 9/203. Remaining: 194\n",
      "Created and saved graph 10/203. Remaining: 193\n",
      "Created and saved graph 11/203. Remaining: 192\n",
      "Created and saved graph 12/203. Remaining: 191\n",
      "Created and saved graph 13/203. Remaining: 190\n",
      "Created and saved graph 14/203. Remaining: 189\n",
      "Created and saved graph 15/203. Remaining: 188\n",
      "Created and saved graph 16/203. Remaining: 187\n",
      "Created and saved graph 17/203. Remaining: 186\n",
      "Created and saved graph 18/203. Remaining: 185\n",
      "Created and saved graph 19/203. Remaining: 184\n",
      "Created and saved graph 20/203. Remaining: 183\n",
      "Created and saved graph 21/203. Remaining: 182\n",
      "Created and saved graph 22/203. Remaining: 181\n",
      "Created and saved graph 23/203. Remaining: 180\n",
      "Created and saved graph 24/203. Remaining: 179\n",
      "Created and saved graph 25/203. Remaining: 178\n",
      "Created and saved graph 26/203. Remaining: 177\n",
      "Created and saved graph 27/203. Remaining: 176\n",
      "Created and saved graph 28/203. Remaining: 175\n",
      "Created and saved graph 29/203. Remaining: 174\n",
      "Created and saved graph 30/203. Remaining: 173\n",
      "Created and saved graph 31/203. Remaining: 172\n",
      "Created and saved graph 32/203. Remaining: 171\n",
      "Created and saved graph 33/203. Remaining: 170\n",
      "Created and saved graph 34/203. Remaining: 169\n",
      "Created and saved graph 35/203. Remaining: 168\n",
      "Created and saved graph 36/203. Remaining: 167\n",
      "Created and saved graph 37/203. Remaining: 166\n",
      "Created and saved graph 38/203. Remaining: 165\n",
      "Created and saved graph 39/203. Remaining: 164\n",
      "Created and saved graph 40/203. Remaining: 163\n",
      "Created and saved graph 41/203. Remaining: 162\n",
      "Created and saved graph 42/203. Remaining: 161\n",
      "Created and saved graph 43/203. Remaining: 160\n",
      "Created and saved graph 44/203. Remaining: 159\n",
      "Created and saved graph 45/203. Remaining: 158\n",
      "Created and saved graph 46/203. Remaining: 157\n",
      "Created and saved graph 47/203. Remaining: 156\n",
      "Created and saved graph 48/203. Remaining: 155\n",
      "Created and saved graph 49/203. Remaining: 154\n",
      "Created and saved graph 50/203. Remaining: 153\n",
      "Created and saved graph 51/203. Remaining: 152\n",
      "Created and saved graph 52/203. Remaining: 151\n",
      "Created and saved graph 53/203. Remaining: 150\n",
      "Created and saved graph 54/203. Remaining: 149\n",
      "Created and saved graph 55/203. Remaining: 148\n",
      "Created and saved graph 56/203. Remaining: 147\n",
      "Created and saved graph 57/203. Remaining: 146\n",
      "Created and saved graph 58/203. Remaining: 145\n",
      "Created and saved graph 59/203. Remaining: 144\n",
      "Created and saved graph 60/203. Remaining: 143\n",
      "Created and saved graph 61/203. Remaining: 142\n",
      "Created and saved graph 62/203. Remaining: 141\n",
      "Created and saved graph 63/203. Remaining: 140\n",
      "Created and saved graph 64/203. Remaining: 139\n",
      "Created and saved graph 65/203. Remaining: 138\n",
      "Created and saved graph 66/203. Remaining: 137\n",
      "Created and saved graph 67/203. Remaining: 136\n",
      "Created and saved graph 68/203. Remaining: 135\n",
      "Created and saved graph 69/203. Remaining: 134\n",
      "Created and saved graph 70/203. Remaining: 133\n",
      "Created and saved graph 71/203. Remaining: 132\n",
      "Created and saved graph 72/203. Remaining: 131\n",
      "Created and saved graph 73/203. Remaining: 130\n",
      "Created and saved graph 74/203. Remaining: 129\n",
      "Created and saved graph 75/203. Remaining: 128\n",
      "Created and saved graph 76/203. Remaining: 127\n",
      "Created and saved graph 77/203. Remaining: 126\n",
      "Created and saved graph 78/203. Remaining: 125\n",
      "Created and saved graph 79/203. Remaining: 124\n",
      "Created and saved graph 80/203. Remaining: 123\n",
      "Created and saved graph 81/203. Remaining: 122\n",
      "Created and saved graph 82/203. Remaining: 121\n",
      "Created and saved graph 83/203. Remaining: 120\n",
      "Created and saved graph 84/203. Remaining: 119\n",
      "Created and saved graph 85/203. Remaining: 118\n",
      "Created and saved graph 86/203. Remaining: 117\n",
      "Created and saved graph 87/203. Remaining: 116\n",
      "Created and saved graph 88/203. Remaining: 115\n",
      "Created and saved graph 89/203. Remaining: 114\n",
      "Created and saved graph 90/203. Remaining: 113\n",
      "Created and saved graph 91/203. Remaining: 112\n",
      "Created and saved graph 92/203. Remaining: 111\n",
      "Created and saved graph 93/203. Remaining: 110\n",
      "Created and saved graph 94/203. Remaining: 109\n",
      "Created and saved graph 95/203. Remaining: 108\n",
      "Created and saved graph 96/203. Remaining: 107\n",
      "Created and saved graph 97/203. Remaining: 106\n",
      "Created and saved graph 98/203. Remaining: 105\n",
      "Created and saved graph 99/203. Remaining: 104\n",
      "Created and saved graph 100/203. Remaining: 103\n",
      "Created and saved graph 101/203. Remaining: 102\n",
      "Created and saved graph 102/203. Remaining: 101\n",
      "Created and saved graph 103/203. Remaining: 100\n",
      "Created and saved graph 104/203. Remaining: 99\n",
      "Created and saved graph 105/203. Remaining: 98\n",
      "Created and saved graph 106/203. Remaining: 97\n",
      "Created and saved graph 107/203. Remaining: 96\n",
      "Created and saved graph 108/203. Remaining: 95\n",
      "Created and saved graph 109/203. Remaining: 94\n",
      "Created and saved graph 110/203. Remaining: 93\n",
      "Created and saved graph 111/203. Remaining: 92\n",
      "Created and saved graph 112/203. Remaining: 91\n",
      "Created and saved graph 113/203. Remaining: 90\n",
      "Created and saved graph 114/203. Remaining: 89\n",
      "Created and saved graph 115/203. Remaining: 88\n",
      "Created and saved graph 116/203. Remaining: 87\n",
      "Created and saved graph 117/203. Remaining: 86\n",
      "Created and saved graph 118/203. Remaining: 85\n",
      "Created and saved graph 119/203. Remaining: 84\n",
      "Created and saved graph 120/203. Remaining: 83\n",
      "Created and saved graph 121/203. Remaining: 82\n",
      "Created and saved graph 122/203. Remaining: 81\n",
      "Created and saved graph 123/203. Remaining: 80\n",
      "Created and saved graph 124/203. Remaining: 79\n",
      "Created and saved graph 125/203. Remaining: 78\n",
      "Created and saved graph 126/203. Remaining: 77\n",
      "Created and saved graph 127/203. Remaining: 76\n",
      "Created and saved graph 128/203. Remaining: 75\n",
      "Created and saved graph 129/203. Remaining: 74\n",
      "Created and saved graph 130/203. Remaining: 73\n",
      "Created and saved graph 131/203. Remaining: 72\n",
      "Created and saved graph 132/203. Remaining: 71\n",
      "Created and saved graph 133/203. Remaining: 70\n",
      "Created and saved graph 134/203. Remaining: 69\n",
      "Created and saved graph 135/203. Remaining: 68\n",
      "Created and saved graph 136/203. Remaining: 67\n",
      "Created and saved graph 137/203. Remaining: 66\n",
      "Created and saved graph 138/203. Remaining: 65\n",
      "Created and saved graph 139/203. Remaining: 64\n",
      "Created and saved graph 140/203. Remaining: 63\n",
      "Created and saved graph 141/203. Remaining: 62\n",
      "Created and saved graph 142/203. Remaining: 61\n",
      "Created and saved graph 143/203. Remaining: 60\n",
      "Created and saved graph 144/203. Remaining: 59\n",
      "Created and saved graph 145/203. Remaining: 58\n",
      "Created and saved graph 146/203. Remaining: 57\n",
      "Created and saved graph 147/203. Remaining: 56\n",
      "Created and saved graph 148/203. Remaining: 55\n",
      "Created and saved graph 149/203. Remaining: 54\n",
      "Created and saved graph 150/203. Remaining: 53\n",
      "Created and saved graph 151/203. Remaining: 52\n",
      "Created and saved graph 152/203. Remaining: 51\n",
      "Created and saved graph 153/203. Remaining: 50\n",
      "Created and saved graph 154/203. Remaining: 49\n",
      "Created and saved graph 155/203. Remaining: 48\n",
      "Created and saved graph 156/203. Remaining: 47\n",
      "Created and saved graph 157/203. Remaining: 46\n",
      "Created and saved graph 158/203. Remaining: 45\n",
      "Created and saved graph 159/203. Remaining: 44\n",
      "Created and saved graph 160/203. Remaining: 43\n",
      "Created and saved graph 161/203. Remaining: 42\n",
      "Created and saved graph 162/203. Remaining: 41\n",
      "Created and saved graph 163/203. Remaining: 40\n",
      "Created and saved graph 164/203. Remaining: 39\n",
      "Created and saved graph 165/203. Remaining: 38\n",
      "Created and saved graph 166/203. Remaining: 37\n",
      "Created and saved graph 167/203. Remaining: 36\n",
      "Created and saved graph 168/203. Remaining: 35\n",
      "Created and saved graph 169/203. Remaining: 34\n",
      "Created and saved graph 170/203. Remaining: 33\n",
      "Created and saved graph 171/203. Remaining: 32\n",
      "Created and saved graph 172/203. Remaining: 31\n",
      "Created and saved graph 173/203. Remaining: 30\n",
      "Created and saved graph 174/203. Remaining: 29\n",
      "Created and saved graph 175/203. Remaining: 28\n",
      "Created and saved graph 176/203. Remaining: 27\n",
      "Created and saved graph 177/203. Remaining: 26\n",
      "Created and saved graph 178/203. Remaining: 25\n",
      "Created and saved graph 179/203. Remaining: 24\n",
      "Created and saved graph 180/203. Remaining: 23\n",
      "Created and saved graph 181/203. Remaining: 22\n",
      "Created and saved graph 182/203. Remaining: 21\n",
      "Created and saved graph 183/203. Remaining: 20\n",
      "Created and saved graph 184/203. Remaining: 19\n",
      "Created and saved graph 185/203. Remaining: 18\n",
      "Created and saved graph 186/203. Remaining: 17\n",
      "Created and saved graph 187/203. Remaining: 16\n",
      "Created and saved graph 188/203. Remaining: 15\n",
      "Created and saved graph 189/203. Remaining: 14\n",
      "Created and saved graph 190/203. Remaining: 13\n",
      "Created and saved graph 191/203. Remaining: 12\n",
      "Created and saved graph 192/203. Remaining: 11\n",
      "Created and saved graph 193/203. Remaining: 10\n",
      "Created and saved graph 194/203. Remaining: 9\n",
      "Created and saved graph 195/203. Remaining: 8\n",
      "Created and saved graph 196/203. Remaining: 7\n",
      "Created and saved graph 197/203. Remaining: 6\n",
      "Created and saved graph 198/203. Remaining: 5\n",
      "Created and saved graph 199/203. Remaining: 4\n",
      "Created and saved graph 200/203. Remaining: 3\n",
      "Created and saved graph 201/203. Remaining: 2\n",
      "Created and saved graph 202/203. Remaining: 1\n",
      "Created and saved graph 203/203. Remaining: 0\n",
      "Graph for P04629: Data(edge_index=[2, 4666], features=[796, 54], num_nodes=796, x=[796, 54])\n",
      "Graph for P15056: Data(edge_index=[2, 4504], features=[766, 54], num_nodes=766, x=[766, 54])\n",
      "Graph for Q13976: Data(edge_index=[2, 5104], features=[671, 54], num_nodes=671, x=[671, 54])\n",
      "Graph for Q96PF2: Data(edge_index=[2, 2470], features=[358, 54], num_nodes=358, x=[358, 54])\n",
      "Graph for Q05655: Data(edge_index=[2, 4212], features=[676, 54], num_nodes=676, x=[676, 54])\n",
      "Graph for O00311: Data(edge_index=[2, 3536], features=[574, 54], num_nodes=574, x=[574, 54])\n",
      "Graph for P43403: Data(edge_index=[2, 4026], features=[619, 54], num_nodes=619, x=[619, 54])\n",
      "Graph for P54619: Data(edge_index=[2, 2400], features=[331, 54], num_nodes=331, x=[331, 54])\n",
      "Graph for P49674: Data(edge_index=[2, 2594], features=[416, 54], num_nodes=416, x=[416, 54])\n",
      "Graph for Q96RR4: Data(edge_index=[2, 3580], features=[588, 54], num_nodes=588, x=[588, 54])\n",
      "Graph for P36507: Data(edge_index=[2, 2814], features=[400, 54], num_nodes=400, x=[400, 54])\n",
      "Graph for P22612: Data(edge_index=[2, 2606], features=[351, 54], num_nodes=351, x=[351, 54])\n",
      "Graph for P51617: Data(edge_index=[2, 4166], features=[712, 54], num_nodes=712, x=[712, 54])\n",
      "Graph for Q9UGJ0: Data(edge_index=[2, 3412], features=[569, 54], num_nodes=569, x=[569, 54])\n",
      "Graph for Q96GD4: Data(edge_index=[2, 2498], features=[344, 54], num_nodes=344, x=[344, 54])\n",
      "Graph for P17252: Data(edge_index=[2, 4126], features=[672, 54], num_nodes=672, x=[672, 54])\n",
      "Graph for Q9HBY8: Data(edge_index=[2, 2946], features=[427, 54], num_nodes=427, x=[427, 54])\n",
      "Graph for Q13237: Data(edge_index=[2, 5464], features=[762, 54], num_nodes=762, x=[762, 54])\n",
      "Graph for Q9H0K1: Data(edge_index=[2, 4958], features=[926, 54], num_nodes=926, x=[926, 54])\n",
      "Graph for Q13627: Data(edge_index=[2, 5742], features=[763, 54], num_nodes=763, x=[763, 54])\n",
      "Graph for O96013: Data(edge_index=[2, 3444], features=[591, 54], num_nodes=591, x=[591, 54])\n",
      "Graph for Q9H3Y6: Data(edge_index=[2, 3360], features=[488, 54], num_nodes=488, x=[488, 54])\n",
      "Graph for P22607: Data(edge_index=[2, 4782], features=[806, 54], num_nodes=806, x=[806, 54])\n",
      "Graph for P42685: Data(edge_index=[2, 3460], features=[505, 54], num_nodes=505, x=[505, 54])\n",
      "Graph for P07332: Data(edge_index=[2, 5026], features=[822, 54], num_nodes=822, x=[822, 54])\n",
      "Graph for Q15078: Data(edge_index=[2, 1554], features=[307, 54], num_nodes=307, x=[307, 54])\n",
      "Graph for P52333: Data(edge_index=[2, 6958], features=[1124, 54], num_nodes=1124, x=[1124, 54])\n",
      "Graph for Q9UGI9: Data(edge_index=[2, 2994], features=[489, 54], num_nodes=489, x=[489, 54])\n",
      "Graph for O96017: Data(edge_index=[2, 3498], features=[543, 54], num_nodes=543, x=[543, 54])\n",
      "Graph for P42336: Data(edge_index=[2, 7034], features=[1068, 54], num_nodes=1068, x=[1068, 54])\n",
      "Graph for Q14164: Data(edge_index=[2, 4762], features=[716, 54], num_nodes=716, x=[716, 54])\n",
      "Graph for O15530: Data(edge_index=[2, 3698], features=[556, 54], num_nodes=556, x=[556, 54])\n",
      "Graph for P49840: Data(edge_index=[2, 5466], features=[483, 54], num_nodes=483, x=[483, 54])\n",
      "Graph for P51955: Data(edge_index=[2, 2606], features=[445, 54], num_nodes=445, x=[445, 54])\n",
      "Graph for Q13555: Data(edge_index=[2, 3734], features=[558, 54], num_nodes=558, x=[558, 54])\n",
      "Graph for Q15118: Data(edge_index=[2, 2940], features=[436, 54], num_nodes=436, x=[436, 54])\n",
      "Graph for Q13554: Data(edge_index=[2, 4218], features=[666, 54], num_nodes=666, x=[666, 54])\n",
      "Graph for P43405: Data(edge_index=[2, 4106], features=[635, 54], num_nodes=635, x=[635, 54])\n",
      "Graph for P49841: Data(edge_index=[2, 2710], features=[420, 54], num_nodes=420, x=[420, 54])\n",
      "Graph for Q12851: Data(edge_index=[2, 4790], features=[820, 54], num_nodes=820, x=[820, 54])\n",
      "Graph for P07333: Data(edge_index=[2, 5650], features=[972, 54], num_nodes=972, x=[972, 54])\n",
      "Graph for P12931: Data(edge_index=[2, 3518], features=[536, 54], num_nodes=536, x=[536, 54])\n",
      "Graph for Q9P1W9: Data(edge_index=[2, 2290], features=[311, 54], num_nodes=311, x=[311, 54])\n",
      "Graph for P42679: Data(edge_index=[2, 3378], features=[507, 54], num_nodes=507, x=[507, 54])\n",
      "Graph for P15735: Data(edge_index=[2, 2946], features=[406, 54], num_nodes=406, x=[406, 54])\n",
      "Graph for Q16566: Data(edge_index=[2, 3074], features=[473, 54], num_nodes=473, x=[473, 54])\n",
      "Graph for P06239: Data(edge_index=[2, 3434], features=[509, 54], num_nodes=509, x=[509, 54])\n",
      "Graph for Q07912: Data(edge_index=[2, 5650], features=[1038, 54], num_nodes=1038, x=[1038, 54])\n",
      "Graph for P11802: Data(edge_index=[2, 2284], features=[303, 54], num_nodes=303, x=[303, 54])\n",
      "Graph for Q7KZI7: Data(edge_index=[2, 4582], features=[788, 54], num_nodes=788, x=[788, 54])\n",
      "Graph for Q04771: Data(edge_index=[2, 3214], features=[509, 54], num_nodes=509, x=[509, 54])\n",
      "Graph for Q04759: Data(edge_index=[2, 4358], features=[706, 54], num_nodes=706, x=[706, 54])\n",
      "Graph for Q13557: Data(edge_index=[2, 3378], features=[499, 54], num_nodes=499, x=[499, 54])\n",
      "Graph for O00329: Data(edge_index=[2, 6940], features=[1044, 54], num_nodes=1044, x=[1044, 54])\n",
      "Graph for P51957: Data(edge_index=[2, 4740], features=[841, 54], num_nodes=841, x=[841, 54])\n",
      "Graph for P23443: Data(edge_index=[2, 3274], features=[525, 54], num_nodes=525, x=[525, 54])\n",
      "Graph for Q14012: Data(edge_index=[2, 2544], features=[370, 54], num_nodes=370, x=[370, 54])\n",
      "Graph for Q13153: Data(edge_index=[2, 3532], features=[545, 54], num_nodes=545, x=[545, 54])\n",
      "Graph for Q9HCP0: Data(edge_index=[2, 2748], features=[422, 54], num_nodes=422, x=[422, 54])\n",
      "Graph for Q02763: Data(edge_index=[2, 6216], features=[1124, 54], num_nodes=1124, x=[1124, 54])\n",
      "Graph for P29323: Data(edge_index=[2, 6330], features=[1055, 54], num_nodes=1055, x=[1055, 54])\n",
      "Graph for P07948: Data(edge_index=[2, 3530], features=[512, 54], num_nodes=512, x=[512, 54])\n",
      "Graph for P48736: Data(edge_index=[2, 7038], features=[1102, 54], num_nodes=1102, x=[1102, 54])\n",
      "Graph for P19784: Data(edge_index=[2, 2558], features=[350, 54], num_nodes=350, x=[350, 54])\n",
      "Graph for P53667: Data(edge_index=[2, 3980], features=[647, 54], num_nodes=647, x=[647, 54])\n",
      "Graph for P54646: Data(edge_index=[2, 3750], features=[552, 54], num_nodes=552, x=[552, 54])\n",
      "Graph for Q14289: Data(edge_index=[2, 6052], features=[1009, 54], num_nodes=1009, x=[1009, 54])\n",
      "Graph for Q9P289: Data(edge_index=[2, 2956], features=[416, 54], num_nodes=416, x=[416, 54])\n",
      "Graph for Q15418: Data(edge_index=[2, 4740], features=[735, 54], num_nodes=735, x=[735, 54])\n",
      "Graph for P07949: Data(edge_index=[2, 5782], features=[1114, 54], num_nodes=1114, x=[1114, 54])\n",
      "Graph for P24941: Data(edge_index=[2, 2252], features=[298, 54], num_nodes=298, x=[298, 54])\n",
      "Graph for Q8TDC3: Data(edge_index=[2, 4668], features=[778, 54], num_nodes=778, x=[778, 54])\n",
      "Graph for Q9UBE8: Data(edge_index=[2, 3510], features=[527, 54], num_nodes=527, x=[527, 54])\n",
      "Graph for Q8IU85: Data(edge_index=[2, 2670], features=[385, 54], num_nodes=385, x=[385, 54])\n",
      "Graph for P09769: Data(edge_index=[2, 3532], features=[529, 54], num_nodes=529, x=[529, 54])\n",
      "Graph for P31751: Data(edge_index=[2, 3258], features=[481, 54], num_nodes=481, x=[481, 54])\n",
      "Graph for P45984: Data(edge_index=[2, 2880], features=[424, 54], num_nodes=424, x=[424, 54])\n",
      "Graph for P34947: Data(edge_index=[2, 4080], features=[590, 54], num_nodes=590, x=[590, 54])\n",
      "Graph for P49760: Data(edge_index=[2, 3014], features=[499, 54], num_nodes=499, x=[499, 54])\n",
      "Graph for Q8IW41: Data(edge_index=[2, 3184], features=[473, 54], num_nodes=473, x=[473, 54])\n",
      "Graph for Q16512: Data(edge_index=[2, 5702], features=[942, 54], num_nodes=942, x=[942, 54])\n",
      "Graph for Q9UBS0: Data(edge_index=[2, 3018], features=[482, 54], num_nodes=482, x=[482, 54])\n",
      "Graph for P21802: Data(edge_index=[2, 4926], features=[821, 54], num_nodes=821, x=[821, 54])\n",
      "Graph for P48730: Data(edge_index=[2, 2616], features=[415, 54], num_nodes=415, x=[415, 54])\n",
      "Graph for Q9UHD2: Data(edge_index=[2, 4884], features=[729, 54], num_nodes=729, x=[729, 54])\n",
      "Graph for Q05397: Data(edge_index=[2, 6376], features=[1052, 54], num_nodes=1052, x=[1052, 54])\n",
      "Graph for P78368: Data(edge_index=[2, 2750], features=[415, 54], num_nodes=415, x=[415, 54])\n",
      "Graph for P11309: Data(edge_index=[2, 2550], features=[404, 54], num_nodes=404, x=[404, 54])\n",
      "Graph for Q08881: Data(edge_index=[2, 4008], features=[620, 54], num_nodes=620, x=[620, 54])\n",
      "Graph for P50750: Data(edge_index=[2, 2638], features=[372, 54], num_nodes=372, x=[372, 54])\n",
      "Graph for P09619: Data(edge_index=[2, 6108], features=[1106, 54], num_nodes=1106, x=[1106, 54])\n",
      "Graph for Q9HAZ1: Data(edge_index=[2, 3010], features=[481, 54], num_nodes=481, x=[481, 54])\n",
      "Graph for Q16513: Data(edge_index=[2, 5866], features=[984, 54], num_nodes=984, x=[984, 54])\n",
      "Graph for P49759: Data(edge_index=[2, 3030], features=[484, 54], num_nodes=484, x=[484, 54])\n",
      "Graph for Q16288: Data(edge_index=[2, 4982], features=[839, 54], num_nodes=839, x=[839, 54])\n",
      "Graph for Q16539: Data(edge_index=[2, 2668], features=[360, 54], num_nodes=360, x=[360, 54])\n",
      "Graph for Q9Y478: Data(edge_index=[2, 1642], features=[270, 54], num_nodes=270, x=[270, 54])\n",
      "Graph for O60285: Data(edge_index=[2, 3992], features=[661, 54], num_nodes=661, x=[661, 54])\n",
      "Graph for Q13131: Data(edge_index=[2, 3774], features=[559, 54], num_nodes=559, x=[559, 54])\n",
      "Graph for P04049: Data(edge_index=[2, 3938], features=[648, 54], num_nodes=648, x=[648, 54])\n",
      "Graph for P45983: Data(edge_index=[2, 2880], features=[427, 54], num_nodes=427, x=[427, 54])\n",
      "Graph for Q9BWU1: Data(edge_index=[2, 3326], features=[502, 54], num_nodes=502, x=[502, 54])\n",
      "Graph for Q9Y243: Data(edge_index=[2, 3262], features=[479, 54], num_nodes=479, x=[479, 54])\n",
      "Graph for O14920: Data(edge_index=[2, 4474], features=[756, 54], num_nodes=756, x=[756, 54])\n",
      "Graph for O75676: Data(edge_index=[2, 4886], features=[772, 54], num_nodes=772, x=[772, 54])\n",
      "Graph for Q9NWZ3: Data(edge_index=[2, 2982], features=[460, 54], num_nodes=460, x=[460, 54])\n",
      "Graph for P67870: Data(edge_index=[2, 1306], features=[215, 54], num_nodes=215, x=[215, 54])\n",
      "Graph for P06493: Data(edge_index=[2, 2302], features=[297, 54], num_nodes=297, x=[297, 54])\n",
      "Graph for P52564: Data(edge_index=[2, 2456], features=[334, 54], num_nodes=334, x=[334, 54])\n",
      "Graph for Q96L34: Data(edge_index=[2, 4432], features=[752, 54], num_nodes=752, x=[752, 54])\n",
      "Graph for Q9HBH9: Data(edge_index=[2, 3092], features=[465, 54], num_nodes=465, x=[465, 54])\n",
      "Graph for P27448: Data(edge_index=[2, 4470], features=[753, 54], num_nodes=753, x=[753, 54])\n",
      "Graph for P29317: Data(edge_index=[2, 5882], features=[976, 54], num_nodes=976, x=[976, 54])\n",
      "Graph for Q13882: Data(edge_index=[2, 3246], features=[451, 54], num_nodes=451, x=[451, 54])\n",
      "Graph for Q86V86: Data(edge_index=[2, 2378], features=[326, 54], num_nodes=326, x=[326, 54])\n",
      "Graph for P30530: Data(edge_index=[2, 4986], features=[894, 54], num_nodes=894, x=[894, 54])\n",
      "Graph for Q15759: Data(edge_index=[2, 2666], features=[364, 54], num_nodes=364, x=[364, 54])\n",
      "Graph for Q9Y463: Data(edge_index=[2, 3916], features=[629, 54], num_nodes=629, x=[629, 54])\n",
      "Graph for Q9NYL2: Data(edge_index=[2, 4684], features=[800, 54], num_nodes=800, x=[800, 54])\n",
      "Graph for P06241: Data(edge_index=[2, 3618], features=[537, 54], num_nodes=537, x=[537, 54])\n",
      "Graph for Q9HC98: Data(edge_index=[2, 2352], features=[313, 54], num_nodes=313, x=[313, 54])\n",
      "Graph for P41240: Data(edge_index=[2, 3204], features=[450, 54], num_nodes=450, x=[450, 54])\n",
      "Graph for Q9BUB5: Data(edge_index=[2, 3074], features=[465, 54], num_nodes=465, x=[465, 54])\n",
      "Graph for P48729: Data(edge_index=[2, 2344], features=[337, 54], num_nodes=337, x=[337, 54])\n",
      "Graph for P80192: Data(edge_index=[2, 5414], features=[1104, 54], num_nodes=1104, x=[1104, 54])\n",
      "Graph for Q9UBF8: Data(edge_index=[2, 5038], features=[816, 54], num_nodes=816, x=[816, 54])\n",
      "Graph for Q9NR20: Data(edge_index=[2, 3292], features=[520, 54], num_nodes=520, x=[520, 54])\n",
      "Graph for Q9Y6M4: Data(edge_index=[2, 2852], features=[447, 54], num_nodes=447, x=[447, 54])\n",
      "Graph for P05771: Data(edge_index=[2, 4164], features=[671, 54], num_nodes=671, x=[671, 54])\n",
      "Graph for Q9H4B4: Data(edge_index=[2, 4440], features=[646, 54], num_nodes=646, x=[646, 54])\n",
      "Graph for Q02750: Data(edge_index=[2, 2812], features=[393, 54], num_nodes=393, x=[393, 54])\n",
      "Graph for O60674: Data(edge_index=[2, 7058], features=[1132, 54], num_nodes=1132, x=[1132, 54])\n",
      "Graph for P07947: Data(edge_index=[2, 3586], features=[543, 54], num_nodes=543, x=[543, 54])\n",
      "Graph for Q02156: Data(edge_index=[2, 4646], features=[737, 54], num_nodes=737, x=[737, 54])\n",
      "Graph for O43293: Data(edge_index=[2, 2602], features=[454, 54], num_nodes=454, x=[454, 54])\n",
      "Graph for P31749: Data(edge_index=[2, 3300], features=[480, 54], num_nodes=480, x=[480, 54])\n",
      "Graph for P41279: Data(edge_index=[2, 3008], features=[467, 54], num_nodes=467, x=[467, 54])\n",
      "Graph for Q16644: Data(edge_index=[2, 2692], features=[382, 54], num_nodes=382, x=[382, 54])\n",
      "Graph for Q02779: Data(edge_index=[2, 4986], features=[954, 54], num_nodes=954, x=[954, 54])\n",
      "Graph for P17612: Data(edge_index=[2, 2604], features=[351, 54], num_nodes=351, x=[351, 54])\n",
      "Graph for P30291: Data(edge_index=[2, 3808], features=[646, 54], num_nodes=646, x=[646, 54])\n",
      "Graph for Q14680: Data(edge_index=[2, 4026], features=[651, 54], num_nodes=651, x=[651, 54])\n",
      "Graph for P51451: Data(edge_index=[2, 3478], features=[505, 54], num_nodes=505, x=[505, 54])\n",
      "Graph for P50613: Data(edge_index=[2, 2434], features=[346, 54], num_nodes=346, x=[346, 54])\n",
      "Graph for Q96KB5: Data(edge_index=[2, 2376], features=[322, 54], num_nodes=322, x=[322, 54])\n",
      "Graph for O00418: Data(edge_index=[2, 4304], features=[725, 54], num_nodes=725, x=[725, 54])\n",
      "Graph for O00141: Data(edge_index=[2, 3074], features=[431, 54], num_nodes=431, x=[431, 54])\n",
      "Graph for P68400: Data(edge_index=[2, 2678], features=[391, 54], num_nodes=391, x=[391, 54])\n",
      "Graph for P22455: Data(edge_index=[2, 4630], features=[802, 54], num_nodes=802, x=[802, 54])\n",
      "Graph for P22694: Data(edge_index=[2, 2620], features=[351, 54], num_nodes=351, x=[351, 54])\n",
      "Graph for P51812: Data(edge_index=[2, 4850], features=[740, 54], num_nodes=740, x=[740, 54])\n",
      "Graph for P53778: Data(edge_index=[2, 2698], features=[367, 54], num_nodes=367, x=[367, 54])\n",
      "Graph for Q16620: Data(edge_index=[2, 4838], features=[822, 54], num_nodes=822, x=[822, 54])\n",
      "Graph for P10721: Data(edge_index=[2, 5626], features=[976, 54], num_nodes=976, x=[976, 54])\n",
      "Graph for O15264: Data(edge_index=[2, 2662], features=[365, 54], num_nodes=365, x=[365, 54])\n",
      "Graph for Q13177: Data(edge_index=[2, 3446], features=[524, 54], num_nodes=524, x=[524, 54])\n",
      "Graph for P42338: Data(edge_index=[2, 7122], features=[1070, 54], num_nodes=1070, x=[1070, 54])\n",
      "Graph for Q13188: Data(edge_index=[2, 3184], features=[491, 54], num_nodes=491, x=[491, 54])\n",
      "Graph for P53779: Data(edge_index=[2, 3100], features=[464, 54], num_nodes=464, x=[464, 54])\n",
      "Graph for Q7L7X3: Data(edge_index=[2, 5554], features=[1001, 54], num_nodes=1001, x=[1001, 54])\n",
      "Graph for P51813: Data(edge_index=[2, 4210], features=[675, 54], num_nodes=675, x=[675, 54])\n",
      "Graph for P36888: Data(edge_index=[2, 5784], features=[993, 54], num_nodes=993, x=[993, 54])\n",
      "Graph for O14965: Data(edge_index=[2, 2594], features=[403, 54], num_nodes=403, x=[403, 54])\n",
      "Graph for P29376: Data(edge_index=[2, 5374], features=[864, 54], num_nodes=864, x=[864, 54])\n",
      "Graph for Q06418: Data(edge_index=[2, 5038], features=[890, 54], num_nodes=890, x=[890, 54])\n",
      "Graph for P00519: Data(edge_index=[2, 6214], features=[1130, 54], num_nodes=1130, x=[1130, 54])\n",
      "Graph for P23458: Data(edge_index=[2, 7048], features=[1154, 54], num_nodes=1154, x=[1154, 54])\n",
      "Graph for Q9BXA7: Data(edge_index=[2, 2566], features=[367, 54], num_nodes=367, x=[367, 54])\n",
      "Graph for O14757: Data(edge_index=[2, 3368], features=[476, 54], num_nodes=476, x=[476, 54])\n",
      "Graph for O15111: Data(edge_index=[2, 4820], features=[745, 54], num_nodes=745, x=[745, 54])\n",
      "Graph for Q9UEE5: Data(edge_index=[2, 2844], features=[414, 54], num_nodes=414, x=[414, 54])\n",
      "Graph for Q9BZL6: Data(edge_index=[2, 5110], features=[878, 54], num_nodes=878, x=[878, 54])\n",
      "Graph for P28482: Data(edge_index=[2, 2648], features=[360, 54], num_nodes=360, x=[360, 54])\n",
      "Graph for O94806: Data(edge_index=[2, 5120], features=[890, 54], num_nodes=890, x=[890, 54])\n",
      "Graph for P24723: Data(edge_index=[2, 4290], features=[683, 54], num_nodes=683, x=[683, 54])\n",
      "Graph for Q8NE63: Data(edge_index=[2, 3706], features=[616, 54], num_nodes=616, x=[616, 54])\n",
      "Graph for P49336: Data(edge_index=[2, 2908], features=[464, 54], num_nodes=464, x=[464, 54])\n",
      "Graph for O15075: Data(edge_index=[2, 4364], features=[740, 54], num_nodes=740, x=[740, 54])\n",
      "Graph for O00444: Data(edge_index=[2, 5264], features=[970, 54], num_nodes=970, x=[970, 54])\n",
      "Graph for Q00534: Data(edge_index=[2, 2382], features=[326, 54], num_nodes=326, x=[326, 54])\n",
      "Graph for Q15139: Data(edge_index=[2, 5030], features=[912, 54], num_nodes=912, x=[912, 54])\n",
      "Graph for Q00535: Data(edge_index=[2, 2200], features=[292, 54], num_nodes=292, x=[292, 54])\n",
      "Graph for P11362: Data(edge_index=[2, 4914], features=[822, 54], num_nodes=822, x=[822, 54])\n",
      "Graph for Q06187: Data(edge_index=[2, 4152], features=[659, 54], num_nodes=659, x=[659, 54])\n",
      "Graph for P05129: Data(edge_index=[2, 4224], features=[697, 54], num_nodes=697, x=[697, 54])\n",
      "Graph for P41743: Data(edge_index=[2, 3962], features=[596, 54], num_nodes=596, x=[596, 54])\n",
      "Graph for O75582: Data(edge_index=[2, 5034], features=[802, 54], num_nodes=802, x=[802, 54])\n",
      "Graph for O43781: Data(edge_index=[2, 3700], features=[588, 54], num_nodes=588, x=[588, 54])\n",
      "Graph for Q9Y4K4: Data(edge_index=[2, 4918], features=[846, 54], num_nodes=846, x=[846, 54])\n",
      "Graph for Q16584: Data(edge_index=[2, 4596], features=[847, 54], num_nodes=847, x=[847, 54])\n",
      "Graph for P51817: Data(edge_index=[2, 2558], features=[358, 54], num_nodes=358, x=[358, 54])\n",
      "Graph for Q96SB4: Data(edge_index=[2, 4052], features=[655, 54], num_nodes=655, x=[655, 54])\n",
      "Graph for Q05513: Data(edge_index=[2, 3862], features=[592, 54], num_nodes=592, x=[592, 54])\n",
      "Graph for P54760: Data(edge_index=[2, 6108], features=[987, 54], num_nodes=987, x=[987, 54])\n",
      "Graph for P27361: Data(edge_index=[2, 2710], features=[379, 54], num_nodes=379, x=[379, 54])\n",
      "Graph for Q12866: Data(edge_index=[2, 5738], features=[999, 54], num_nodes=999, x=[999, 54])\n",
      "Graph for P08631: Data(edge_index=[2, 3488], features=[526, 54], num_nodes=526, x=[526, 54])\n",
      "Graph for P49137: Data(edge_index=[2, 2730], features=[400, 54], num_nodes=400, x=[400, 54])\n",
      "Graph for Q9UQM7: Data(edge_index=[2, 3362], features=[478, 54], num_nodes=478, x=[478, 54])\n",
      "Graph for P53350: Data(edge_index=[2, 4176], features=[603, 54], num_nodes=603, x=[603, 54])\n",
      "Graph for O43741: Data(edge_index=[2, 1618], features=[272, 54], num_nodes=272, x=[272, 54])\n",
      "Graph for P16234: Data(edge_index=[2, 6042], features=[1089, 54], num_nodes=1089, x=[1089, 54])\n",
      "Graph for P16591: Data(edge_index=[2, 5222], features=[822, 54], num_nodes=822, x=[822, 54])\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import torch\n",
    "import torch_geometric.data as DATA\n",
    "from torch_geometric.utils import from_networkx\n",
    "import networkx as nx\n",
    "\n",
    "# Residue properties for feature generation\n",
    "res_weight_table = {'A': 71.08, 'C': 103.15, 'D': 115.09, 'E': 129.12, 'F': 147.18, 'G': 57.05, 'H': 137.14,\n",
    "                    'I': 113.16, 'K': 128.18, 'L': 113.16, 'M': 131.20, 'N': 114.11, 'P': 97.12, 'Q': 128.13,\n",
    "                    'R': 156.19, 'S': 87.08, 'T': 101.11, 'V': 99.13, 'W': 186.22, 'Y': 163.18 , 'X': 0 }\n",
    "\n",
    "res_pka_table = {'A': 2.34, 'C': 1.96, 'D': 1.88, 'E': 2.19, 'F': 1.83, 'G': 2.34, 'H': 1.82, 'I': 2.36,\n",
    "                 'K': 2.18, 'L': 2.36, 'M': 2.28, 'N': 2.02, 'P': 1.99, 'Q': 2.17, 'R': 2.17, 'S': 2.21,\n",
    "                 'T': 2.09, 'V': 2.32, 'W': 2.83, 'Y': 2.32}\n",
    "\n",
    "res_pkb_table = {'A': 9.69, 'C': 10.28, 'D': 9.60, 'E': 9.67, 'F': 9.13, 'G': 9.60, 'H': 9.17,\n",
    "                 'I': 9.60, 'K': 8.95, 'L': 9.60, 'M': 9.21, 'N': 8.80, 'P': 10.60, 'Q': 9.13,\n",
    "                 'R': 9.04, 'S': 9.15, 'T': 9.10, 'V': 9.62, 'W': 9.39, 'Y': 9.62}\n",
    "\n",
    "res_pkx_table = {'A': 0.00, 'C': 8.18, 'D': 3.65, 'E': 4.25, 'F': 0.00, 'G': 0, 'H': 6.00,\n",
    "                 'I': 0.00, 'K': 10.53, 'L': 0.00, 'M': 0.00, 'N': 0.00, 'P': 0.00, 'Q': 0.00,\n",
    "                 'R': 12.48, 'S': 0.00, 'T': 0.00, 'V': 0.00, 'W': 0.00, 'Y': 0.00}\n",
    "\n",
    "res_pl_table = {'A': 6.00, 'C': 5.07, 'D': 2.77, 'E': 3.22, 'F': 5.48, 'G': 5.97, 'H': 7.59,\n",
    "                'I': 6.02, 'K': 9.74, 'L': 5.98, 'M': 5.74, 'N': 5.41, 'P': 6.30, 'Q': 5.65,\n",
    "                'R': 10.76, 'S': 5.68, 'T': 5.60, 'V': 5.96, 'W': 5.89, 'Y': 5.96}\n",
    "\n",
    "res_hydrophobic_ph2_table = {'A': 47, 'C': 52, 'D': -18, 'E': 8, 'F': 92, 'G': 0, 'H': -42, 'I': 100,\n",
    "                             'K': -37, 'L': 100, 'M': 74, 'N': -41, 'P': -46, 'Q': -18, 'R': -26, 'S': -7,\n",
    "                             'T': 13, 'V': 79, 'W': 84, 'Y': 49}\n",
    "\n",
    "res_hydrophobic_ph7_table = {'A': 41, 'C': 49, 'D': -55, 'E': -31, 'F': 100, 'G': 0, 'H': 8, 'I': 99,\n",
    "                             'K': -23, 'L': 97, 'M': 74, 'N': -28, 'P': -46, 'Q': -10, 'R': -14, 'S': -5,\n",
    "                             'T': 13, 'V': 76, 'W': 97, 'Y': 63}\n",
    "\n",
    "# Residue group properties\n",
    "aliphatic = ['A', 'I', 'L', 'M', 'V']\n",
    "aromatic = ['F', 'W', 'Y']\n",
    "polar_neutral = ['C', 'N', 'Q', 'S', 'T']\n",
    "acidic_charged = ['D', 'E']\n",
    "basic_charged = ['H', 'K', 'R']\n",
    "\n",
    "# Function to calculate PSSM from an alignment file and the protein sequence\n",
    "def PSSM_calculation(aln_file, pro_seq):\n",
    "    \"\"\"\n",
    "    Calculates the PSSM matrix from the alignment file and protein sequence, ignoring gaps in alignment.\n",
    "    \n",
    "    Parameters:\n",
    "    aln_file (str): Path to the alignment file (.aln)\n",
    "    pro_seq (str): The corresponding protein sequence\n",
    "    \n",
    "    Returns:\n",
    "    numpy array: PSSM matrix (Nx21 where N is the length of the protein sequence)\n",
    "    \"\"\"\n",
    "    pfm_mat = np.zeros((len(res_weight_table), len(pro_seq)))  # Frequency matrix initialization\n",
    "    with open(aln_file, 'r') as f:\n",
    "        lines = [line.strip() for line in f.readlines() if line.strip()]  # Read non-empty lines\n",
    "        line_count = len(lines)  # Number of sequences\n",
    "\n",
    "        for line in lines:\n",
    "            if len(line) != len(pro_seq):  # Ignore sequences that don't match the expected length\n",
    "                print(f\"Skipping line with mismatched length: {len(line)} vs expected {len(pro_seq)}\")\n",
    "                continue\n",
    "            count = 0\n",
    "            for res in line:\n",
    "                if res in res_weight_table:\n",
    "                    pfm_mat[list(res_weight_table.keys()).index(res), count] += 1\n",
    "                count += 1\n",
    "\n",
    "    # Convert PFM to PPM using pseudocounts\n",
    "    pseudocount = 0.8\n",
    "    ppm_mat = (pfm_mat + pseudocount / len(res_weight_table)) / (line_count + pseudocount)\n",
    "\n",
    "    # Ensure PSSM matrix has correct dimensions (length of sequence, 21 features)\n",
    "    if ppm_mat.shape[0] != len(pro_seq):\n",
    "        ppm_mat = ppm_mat.T  # Transpose if necessary\n",
    "\n",
    "    return ppm_mat  # The PPM is used as the PSSM matrix in this case\n",
    "\n",
    "\n",
    "\n",
    "# Function to generate the full node feature vector for each residue\n",
    "\n",
    "def residue_features(residue, pssm_vector):\n",
    "    \"\"\"\n",
    "    Generates a feature vector for each residue (amino acid), including PSSM and other biochemical properties.\n",
    "\n",
    "    Parameters:\n",
    "    residue (str): The single-letter amino acid code (e.g., 'A', 'C').\n",
    "    pssm_vector (numpy array): The PSSM vector for this residue (21-dimensional).\n",
    "\n",
    "    Returns:\n",
    "    numpy array: A 54-dimensional feature vector for the residue.\n",
    "    \"\"\"\n",
    "    # One-hot encoding of the residue (21-dimensional)\n",
    "    one_hot = [1 if residue == r else 0 for r in res_weight_table.keys()]\n",
    "\n",
    "    # Residue group properties (binary flags)\n",
    "    aliphatic_flag = [1] if residue in aliphatic else [0]  # 1 feature\n",
    "    aromatic_flag = [1] if residue in aromatic else [0]  # 1 feature\n",
    "    polar_neutral_flag = [1] if residue in polar_neutral else [0]  # 1 feature\n",
    "    acidic_charged_flag = [1] if residue in acidic_charged else [0]  # 1 feature\n",
    "    basic_charged_flag = [1] if residue in basic_charged else [0]  # 1 feature\n",
    "\n",
    "    # Chemical properties (single scalar values)\n",
    "    weight = [res_weight_table.get(residue, 0)]  # Molecular weight (1 feature)\n",
    "    pka = [res_pka_table.get(residue, 0)]  # pKa of -COOH (1 feature)\n",
    "    pkb = [res_pkb_table.get(residue, 0)]  # pKb of -NH3 (1 feature)\n",
    "    pkx = [res_pkx_table.get(residue, 0)]  # pKa of side chain (1 feature)\n",
    "    pi = [res_pl_table.get(residue, 0)]  # Isoelectric point (pI) (1 feature)\n",
    "    hydrophobicity_ph2 = [res_hydrophobic_ph2_table.get(residue, 0)]  # Hydrophobicity at pH 2 (1 feature)\n",
    "    hydrophobicity_ph7 = [res_hydrophobic_ph7_table.get(residue, 0)]  # Hydrophobicity at pH 7 (1 feature)\n",
    "\n",
    "    # Combine all features into a single vector (54-dimensional)\n",
    "    feature_vector = (\n",
    "        one_hot +  # 21 features\n",
    "        list(pssm_vector) +  # PSSM vector (21-dimensional)\n",
    "        aliphatic_flag +  # 1 feature\n",
    "        aromatic_flag +  # 1 feature\n",
    "        polar_neutral_flag +  # 1 feature\n",
    "        acidic_charged_flag +  # 1 feature\n",
    "        basic_charged_flag +  # 1 feature\n",
    "        weight +  # 1 feature\n",
    "        pka +  # 1 feature\n",
    "        pkb +  # 1 feature\n",
    "        pkx +  # 1 feature\n",
    "        pi +  # 1 feature\n",
    "        hydrophobicity_ph2 +  # 1 feature\n",
    "        hydrophobicity_ph7  # 1 feature\n",
    "    )\n",
    "\n",
    "    return np.array(feature_vector)\n",
    "\n",
    "\n",
    "\n",
    "# Function to create the graph from the contact map and node features\n",
    "def create_graph_from_contact_map(protein_sequence, pssm_matrix, contact_map):\n",
    "    \"\"\"\n",
    "    Creates a graph from a protein contact map, adding node features based on residue properties.\n",
    "\n",
    "    Parameters:\n",
    "    protein_sequence (str): The amino acid sequence of the protein.\n",
    "    pssm_matrix (numpy array): The PSSM matrix for the protein (Nx21 for N residues).\n",
    "    contact_map (numpy array): The NxN contact map matrix where each entry indicates a contact.\n",
    "\n",
    "    Returns:\n",
    "    torch_geometric.data.Data: The graph data in PyTorch Geometric format.\n",
    "    \"\"\"\n",
    "    num_residues = len(protein_sequence)  # Number of residues in the sequence\n",
    "\n",
    "    # Ensure PSSM matrix matches the sequence length\n",
    "    if len(pssm_matrix) != num_residues:\n",
    "        print(pssm_matrix)\n",
    "        print(pssm_matrix.shape)\n",
    "        raise ValueError(f\"Mismatch between sequence length ({num_residues}) and PSSM matrix size ({len(pssm_matrix)}).\")\n",
    "\n",
    "    # Create a NetworkX graph\n",
    "    G = nx.Graph()\n",
    "\n",
    "    # Add nodes with residue features\n",
    "    for i, residue in enumerate(protein_sequence):\n",
    "        pssm_vector = pssm_matrix[i]  # Get the PSSM vector for residue i\n",
    "        G.add_node(i, features=residue_features(residue, pssm_vector))\n",
    "\n",
    "    # Add edges based on the contact map (threshold: contact_map >= 0.5)\n",
    "    for i in range(num_residues):\n",
    "        for j in range(i + 1, num_residues):\n",
    "            if contact_map[i, j] >= 0.5:  # Threshold for contact\n",
    "                G.add_edge(i, j)\n",
    "\n",
    "    # Convert the NetworkX graph to PyTorch Geometric format\n",
    "    data = from_networkx(G)\n",
    "\n",
    "    # Extract node features into the `x` attribute (used by PyTorch Geometric)\n",
    "    data.x = torch.tensor([G.nodes[i]['features'] for i in G.nodes], dtype=torch.float)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "# Process the valid sequences and corresponding contact maps\n",
    "def process_proteins(sequence_dict, contact_map_folder, aln_folder, save_folder):\n",
    "    \"\"\"\n",
    "    Processes all proteins: loads contact maps, calculates PSSM matrices, creates graphs with node features, \n",
    "    and saves the graphs to files.\n",
    "    A counter tracks how many graphs are created and how many are left.\n",
    "    \n",
    "    Parameters:\n",
    "    sequence_dict (dict): Dictionary containing the valid protein sequences.\n",
    "    contact_map_folder (str): Path to the folder containing the contact maps.\n",
    "    aln_folder (str): Path to the folder containing the alignment files (.aln).\n",
    "    save_folder (str): Path to the folder where the graph files will be saved.\n",
    "    \n",
    "    Returns:\n",
    "    dict: A dictionary where keys are protein IDs and values are graph data objects.\n",
    "    \"\"\"\n",
    "    graph_data_dict = {}\n",
    "\n",
    "    # Ensure the save folder exists\n",
    "    os.makedirs(save_folder, exist_ok=True)\n",
    "\n",
    "    # Get the total number of .npy files (contact maps) in the folder\n",
    "    contact_files = [f for f in os.listdir(contact_map_folder) if f.endswith('.npy')]\n",
    "    total_files = len(contact_files)\n",
    "    completed = 0\n",
    "\n",
    "    # Process each contact map\n",
    "    for contact_file in contact_files:\n",
    "        protein_id = contact_file.split('.npy')[0]\n",
    "\n",
    "        if protein_id in sequence_dict:\n",
    "            # Load the contact map\n",
    "            contact_map_path = os.path.join(contact_map_folder, contact_file)\n",
    "            contact_map = np.load(contact_map_path)\n",
    "\n",
    "            # Retrieve the protein sequence\n",
    "            sequence = sequence_dict[protein_id]\n",
    "\n",
    "            # Calculate the PSSM matrix for this sequence\n",
    "            aln_file = os.path.join(aln_folder, f\"{protein_id}.aln\")\n",
    "            pssm_matrix = PSSM_calculation(aln_file, sequence)\n",
    "\n",
    "            # Create the graph from the contact map and sequence\n",
    "            graph_data = create_graph_from_contact_map(sequence, pssm_matrix, contact_map)\n",
    "\n",
    "            # Store the graph data in the dictionary\n",
    "            graph_data_dict[protein_id] = graph_data\n",
    "\n",
    "            # Save the graph to a file\n",
    "            graph_save_path = os.path.join(save_folder, f\"{protein_id}_graph.pt\")\n",
    "            torch.save(graph_data, graph_save_path)\n",
    "\n",
    "            # Update the counter\n",
    "            completed += 1\n",
    "            print(f\"Created and saved graph {completed}/{total_files}. Remaining: {total_files - completed}\")\n",
    "        else:\n",
    "            print(f\"Warning: No sequence found for protein {protein_id}\")\n",
    "\n",
    "    return graph_data_dict\n",
    "# Example usage\n",
    "contact_map_folder = \"./KibaContactMaps\"\n",
    "aln_folder = \"aln\"\n",
    "save_folder = \"./ProteinGraphs\"\n",
    "\n",
    "# Assuming valid_sequence_dict is already populated\n",
    "graph_data_dict = process_proteins(valid_sequence_dict, contact_map_folder, aln_folder,save_folder)\n",
    "\n",
    "# You can now use graph_data_dict for downstream tasks\n",
    "for protein_id, graph in graph_data_dict.items():\n",
    "    print(f\"Graph for {protein_id}: {graph}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "acce1016-e76f-46e6-ad9a-37b135b43c9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created and saved graph 1/203. Remaining: 202\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 243\u001b[0m\n\u001b[1;32m    240\u001b[0m save_folder \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./ProteinGraphs\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    242\u001b[0m \u001b[38;5;66;03m# Assuming valid_sequence_dict is already populated\u001b[39;00m\n\u001b[0;32m--> 243\u001b[0m graph_data_dict \u001b[38;5;241m=\u001b[39m process_proteins(valid_sequence_dict, contact_map_folder, aln_folder,save_folder)\n\u001b[1;32m    245\u001b[0m \u001b[38;5;66;03m# You can now use graph_data_dict for downstream tasks\u001b[39;00m\n\u001b[1;32m    246\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m protein_id, graph \u001b[38;5;129;01min\u001b[39;00m graph_data_dict\u001b[38;5;241m.\u001b[39mitems():\n",
      "Cell \u001b[0;32mIn[8], line 218\u001b[0m, in \u001b[0;36mprocess_proteins\u001b[0;34m(sequence_dict, contact_map_folder, aln_folder, save_folder)\u001b[0m\n\u001b[1;32m    216\u001b[0m \u001b[38;5;66;03m# Calculate the PSSM matrix for this sequence\u001b[39;00m\n\u001b[1;32m    217\u001b[0m aln_file \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(aln_folder, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mprotein_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.aln\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 218\u001b[0m pssm_matrix \u001b[38;5;241m=\u001b[39m PSSM_calculation(aln_file, sequence)\n\u001b[1;32m    220\u001b[0m \u001b[38;5;66;03m# Create the graph from the contact map and sequence\u001b[39;00m\n\u001b[1;32m    221\u001b[0m graph_data \u001b[38;5;241m=\u001b[39m create_graph_from_contact_map(sequence, pssm_matrix, contact_map)\n",
      "Cell \u001b[0;32mIn[8], line 68\u001b[0m, in \u001b[0;36mPSSM_calculation\u001b[0;34m(aln_file, pro_seq)\u001b[0m\n\u001b[1;32m     66\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m line:\n\u001b[1;32m     67\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m res_weight_table:\n\u001b[0;32m---> 68\u001b[0m                 pfm_mat[\u001b[38;5;28mlist\u001b[39m(res_weight_table\u001b[38;5;241m.\u001b[39mkeys())\u001b[38;5;241m.\u001b[39mindex(res), count] \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m     69\u001b[0m             count \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;66;03m# Convert PFM to PPM using pseudocounts\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import torch\n",
    "import torch_geometric.data as DATA\n",
    "from torch_geometric.utils import from_networkx\n",
    "import networkx as nx\n",
    "\n",
    "# Residue properties for feature generation\n",
    "res_weight_table = {'A': 71.08, 'C': 103.15, 'D': 115.09, 'E': 129.12, 'F': 147.18, 'G': 57.05, 'H': 137.14,\n",
    "                    'I': 113.16, 'K': 128.18, 'L': 113.16, 'M': 131.20, 'N': 114.11, 'P': 97.12, 'Q': 128.13,\n",
    "                    'R': 156.19, 'S': 87.08, 'T': 101.11, 'V': 99.13, 'W': 186.22, 'Y': 163.18}\n",
    "\n",
    "res_pka_table = {'A': 2.34, 'C': 1.96, 'D': 1.88, 'E': 2.19, 'F': 1.83, 'G': 2.34, 'H': 1.82, 'I': 2.36,\n",
    "                 'K': 2.18, 'L': 2.36, 'M': 2.28, 'N': 2.02, 'P': 1.99, 'Q': 2.17, 'R': 2.17, 'S': 2.21,\n",
    "                 'T': 2.09, 'V': 2.32, 'W': 2.83, 'Y': 2.32}\n",
    "\n",
    "res_pkb_table = {'A': 9.69, 'C': 10.28, 'D': 9.60, 'E': 9.67, 'F': 9.13, 'G': 9.60, 'H': 9.17,\n",
    "                 'I': 9.60, 'K': 8.95, 'L': 9.60, 'M': 9.21, 'N': 8.80, 'P': 10.60, 'Q': 9.13,\n",
    "                 'R': 9.04, 'S': 9.15, 'T': 9.10, 'V': 9.62, 'W': 9.39, 'Y': 9.62}\n",
    "\n",
    "res_pkx_table = {'A': 0.00, 'C': 8.18, 'D': 3.65, 'E': 4.25, 'F': 0.00, 'G': 0, 'H': 6.00,\n",
    "                 'I': 0.00, 'K': 10.53, 'L': 0.00, 'M': 0.00, 'N': 0.00, 'P': 0.00, 'Q': 0.00,\n",
    "                 'R': 12.48, 'S': 0.00, 'T': 0.00, 'V': 0.00, 'W': 0.00, 'Y': 0.00}\n",
    "\n",
    "res_pl_table = {'A': 6.00, 'C': 5.07, 'D': 2.77, 'E': 3.22, 'F': 5.48, 'G': 5.97, 'H': 7.59,\n",
    "                'I': 6.02, 'K': 9.74, 'L': 5.98, 'M': 5.74, 'N': 5.41, 'P': 6.30, 'Q': 5.65,\n",
    "                'R': 10.76, 'S': 5.68, 'T': 5.60, 'V': 5.96, 'W': 5.89, 'Y': 5.96}\n",
    "\n",
    "res_hydrophobic_ph2_table = {'A': 47, 'C': 52, 'D': -18, 'E': 8, 'F': 92, 'G': 0, 'H': -42, 'I': 100,\n",
    "                             'K': -37, 'L': 100, 'M': 74, 'N': -41, 'P': -46, 'Q': -18, 'R': -26, 'S': -7,\n",
    "                             'T': 13, 'V': 79, 'W': 84, 'Y': 49}\n",
    "\n",
    "res_hydrophobic_ph7_table = {'A': 41, 'C': 49, 'D': -55, 'E': -31, 'F': 100, 'G': 0, 'H': 8, 'I': 99,\n",
    "                             'K': -23, 'L': 97, 'M': 74, 'N': -28, 'P': -46, 'Q': -10, 'R': -14, 'S': -5,\n",
    "                             'T': 13, 'V': 76, 'W': 97, 'Y': 63}\n",
    "\n",
    "# Residue group properties\n",
    "aliphatic = ['A', 'I', 'L', 'M', 'V']\n",
    "aromatic = ['F', 'W', 'Y']\n",
    "polar_neutral = ['C', 'N', 'Q', 'S', 'T']\n",
    "acidic_charged = ['D', 'E']\n",
    "basic_charged = ['H', 'K', 'R']\n",
    "\n",
    "# Function to calculate PSSM from an alignment file and the protein sequence\n",
    "def PSSM_calculation(aln_file, pro_seq):\n",
    "    \"\"\"\n",
    "    Calculates the PSSM matrix from the alignment file and protein sequence, ignoring gaps in alignment.\n",
    "    \n",
    "    Parameters:\n",
    "    aln_file (str): Path to the alignment file (.aln)\n",
    "    pro_seq (str): The corresponding protein sequence\n",
    "    \n",
    "    Returns:\n",
    "    numpy array: PSSM matrix (Nx21 where N is the length of the protein sequence)\n",
    "    \"\"\"\n",
    "    pfm_mat = np.zeros((len(res_weight_table), len(pro_seq)))  # Frequency matrix initialization\n",
    "    with open(aln_file, 'r') as f:\n",
    "        lines = [line.strip() for line in f.readlines() if line.strip()]  # Read non-empty lines\n",
    "        line_count = len(lines)  # Number of sequences\n",
    "\n",
    "        for line in lines:\n",
    "            if len(line) != len(pro_seq):  # Ignore sequences that don't match the expected length\n",
    "                print(f\"Skipping line with mismatched length: {len(line)} vs expected {len(pro_seq)}\")\n",
    "                continue\n",
    "            count = 0\n",
    "            for res in line:\n",
    "                if res in res_weight_table:\n",
    "                    pfm_mat[list(res_weight_table.keys()).index(res), count] += 1\n",
    "                count += 1\n",
    "\n",
    "    # Convert PFM to PPM using pseudocounts\n",
    "    pseudocount = 0.8\n",
    "    ppm_mat = (pfm_mat + pseudocount / len(res_weight_table)) / (line_count + pseudocount)\n",
    "\n",
    "    # Ensure PSSM matrix has correct dimensions (length of sequence, 21 features)\n",
    "    if ppm_mat.shape[0] != len(pro_seq):\n",
    "        ppm_mat = ppm_mat.T  # Transpose if necessary\n",
    "\n",
    "    return ppm_mat  # The PPM is used as the PSSM matrix in this case\n",
    "\n",
    "\n",
    "\n",
    "# Function to generate the full node feature vector for each residue\n",
    "def residue_features(residue, pssm_vector):\n",
    "    \"\"\"\n",
    "    Generates a feature vector for each residue (amino acid), including PSSM and other biochemical properties.\n",
    "\n",
    "    Parameters:\n",
    "    residue (str): The single-letter amino acid code (e.g., 'A', 'C').\n",
    "    pssm_vector (numpy array): The PSSM vector for this residue (21-dimensional).\n",
    "\n",
    "    Returns:\n",
    "    numpy array: A 54-dimensional feature vector for the residue.\n",
    "    \"\"\"\n",
    "    # One-hot encoding of the residue (21-dimensional)\n",
    "    one_hot = [1 if residue == r else 0 for r in res_weight_table.keys()]\n",
    "\n",
    "    # Residue group properties (binary flags)\n",
    "    aliphatic_flag = [1] if residue in aliphatic else [0]\n",
    "    aromatic_flag = [1] if residue in aromatic else [0]\n",
    "    polar_neutral_flag = [1] if residue in polar_neutral else [0]\n",
    "    acidic_charged_flag = [1] if residue in acidic_charged else [0]\n",
    "    basic_charged_flag = [1] if residue in basic_charged else [0]\n",
    "\n",
    "    # Chemical properties (single scalar values)\n",
    "    weight = [res_weight_table.get(residue, 0)]  # Molecular weight\n",
    "    pka = [res_pka_table.get(residue, 0)]  # pKa of -COOH\n",
    "    pkb = [res_pkb_table.get(residue, 0)]  # pKb of -NH3\n",
    "    pkx = [res_pkx_table.get(residue, 0)]  # pKa of side chain\n",
    "    pi = [res_pl_table.get(residue, 0)]  # Isoelectric point (pI)\n",
    "    hydrophobicity_ph2 = [res_hydrophobic_ph2_table.get(residue, 0)]  # Hydrophobicity at pH 2\n",
    "    hydrophobicity_ph7 = [res_hydrophobic_ph7_table.get(residue, 0)]  # Hydrophobicity at pH 7\n",
    "\n",
    "    # Combine all features into a single vector (54-dimensional)\n",
    "    feature_vector = (\n",
    "        one_hot +\n",
    "        list(pssm_vector) +  # PSSM vector (21-dimensional)\n",
    "        aliphatic_flag +\n",
    "        aromatic_flag +\n",
    "        polar_neutral_flag +\n",
    "        acidic_charged_flag +\n",
    "        basic_charged_flag +\n",
    "        weight +\n",
    "        pka +\n",
    "        pkb +\n",
    "        pkx +\n",
    "        pi +\n",
    "        hydrophobicity_ph2 +\n",
    "        hydrophobicity_ph7\n",
    "    )\n",
    "    \n",
    "    return np.array(feature_vector)\n",
    "\n",
    "# Function to create the graph from the contact map and node features\n",
    "def create_graph_from_contact_map(protein_sequence, pssm_matrix, contact_map):\n",
    "    \"\"\"\n",
    "    Creates a graph from a protein contact map, adding node features based on residue properties.\n",
    "\n",
    "    Parameters:\n",
    "    protein_sequence (str): The amino acid sequence of the protein.\n",
    "    pssm_matrix (numpy array): The PSSM matrix for the protein (Nx21 for N residues).\n",
    "    contact_map (numpy array): The NxN contact map matrix where each entry indicates a contact.\n",
    "\n",
    "    Returns:\n",
    "    torch_geometric.data.Data: The graph data in PyTorch Geometric format.\n",
    "    \"\"\"\n",
    "    num_residues = len(protein_sequence)  # Number of residues in the sequence\n",
    "\n",
    "    # Ensure PSSM matrix matches the sequence length\n",
    "    if len(pssm_matrix) != num_residues:\n",
    "        print(pssm_matrix)\n",
    "        print(pssm_matrix.shape)\n",
    "        raise ValueError(f\"Mismatch between sequence length ({num_residues}) and PSSM matrix size ({len(pssm_matrix)}).\")\n",
    "\n",
    "    # Create a NetworkX graph\n",
    "    G = nx.Graph()\n",
    "\n",
    "    # Add nodes with residue features\n",
    "    for i, residue in enumerate(protein_sequence):\n",
    "        pssm_vector = pssm_matrix[i]  # Get the PSSM vector for residue i\n",
    "        G.add_node(i, features=residue_features(residue, pssm_vector))\n",
    "\n",
    "    # Add edges based on the contact map (threshold: contact_map >= 0.5)\n",
    "    for i in range(num_residues):\n",
    "        for j in range(i + 1, num_residues):\n",
    "            if contact_map[i, j] >= 0.5:  # Threshold for contact\n",
    "                G.add_edge(i, j)\n",
    "\n",
    "    # Convert the NetworkX graph to PyTorch Geometric format\n",
    "    data = from_networkx(G)\n",
    "\n",
    "    # Extract node features into the `x` attribute (used by PyTorch Geometric)\n",
    "    data.x = torch.tensor([G.nodes[i]['features'] for i in G.nodes], dtype=torch.float)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "# Process the valid sequences and corresponding contact maps\n",
    "def process_proteins(sequence_dict, contact_map_folder, aln_folder, save_folder):\n",
    "    \"\"\"\n",
    "    Processes all proteins: loads contact maps, calculates PSSM matrices, creates graphs with node features, \n",
    "    and saves the graphs to files.\n",
    "    A counter tracks how many graphs are created and how many are left.\n",
    "    \n",
    "    Parameters:\n",
    "    sequence_dict (dict): Dictionary containing the valid protein sequences.\n",
    "    contact_map_folder (str): Path to the folder containing the contact maps.\n",
    "    aln_folder (str): Path to the folder containing the alignment files (.aln).\n",
    "    save_folder (str): Path to the folder where the graph files will be saved.\n",
    "    \n",
    "    Returns:\n",
    "    dict: A dictionary where keys are protein IDs and values are graph data objects.\n",
    "    \"\"\"\n",
    "    graph_data_dict = {}\n",
    "\n",
    "    # Ensure the save folder exists\n",
    "    os.makedirs(save_folder, exist_ok=True)\n",
    "\n",
    "    # Get the total number of .npy files (contact maps) in the folder\n",
    "    contact_files = [f for f in os.listdir(contact_map_folder) if f.endswith('.npy')]\n",
    "    total_files = len(contact_files)\n",
    "    completed = 0\n",
    "\n",
    "    # Process each contact map\n",
    "    for contact_file in contact_files:\n",
    "        protein_id = contact_file.split('.npy')[0]\n",
    "\n",
    "        if protein_id in sequence_dict:\n",
    "            # Load the contact map\n",
    "            contact_map_path = os.path.join(contact_map_folder, contact_file)\n",
    "            contact_map = np.load(contact_map_path)\n",
    "\n",
    "            # Retrieve the protein sequence\n",
    "            sequence = sequence_dict[protein_id]\n",
    "\n",
    "            # Calculate the PSSM matrix for this sequence\n",
    "            aln_file = os.path.join(aln_folder, f\"{protein_id}.aln\")\n",
    "            pssm_matrix = PSSM_calculation(aln_file, sequence)\n",
    "\n",
    "            # Create the graph from the contact map and sequence\n",
    "            graph_data = create_graph_from_contact_map(sequence, pssm_matrix, contact_map)\n",
    "\n",
    "            # Store the graph data in the dictionary\n",
    "            graph_data_dict[protein_id] = graph_data\n",
    "\n",
    "            # Save the graph to a file\n",
    "            graph_save_path = os.path.join(save_folder, f\"{protein_id}_graph.pt\")\n",
    "            torch.save(graph_data, graph_save_path)\n",
    "\n",
    "            # Update the counter\n",
    "            completed += 1\n",
    "            print(f\"Created and saved graph {completed}/{total_files}. Remaining: {total_files - completed}\")\n",
    "        else:\n",
    "            print(f\"Warning: No sequence found for protein {protein_id}\")\n",
    "\n",
    "    return graph_data_dict\n",
    "# Example usage\n",
    "contact_map_folder = \"./KibaContactMaps\"\n",
    "aln_folder = \"aln\"\n",
    "save_folder = \"./ProteinGraphs\"\n",
    "\n",
    "# Assuming valid_sequence_dict is already populated\n",
    "graph_data_dict = process_proteins(valid_sequence_dict, contact_map_folder, aln_folder,save_folder)\n",
    "\n",
    "# You can now use graph_data_dict for downstream tasks\n",
    "for protein_id, graph in graph_data_dict.items():\n",
    "    print(f\"Graph for {protein_id}: {graph}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ba8a192-5296-4420-b277-0e970b89980d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c99eea93-10fe-4fe5-ad78-327fa70f79ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba83762f-672b-4b79-b3fb-3bf2fbded7f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c42bfc3-d5be-4719-ab12-9cfeb8313953",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27fbd0dc-bda7-4444-a6d3-91eb4c026227",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c50394-f80d-46a3-ba72-5bec79b855ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76466df6-6cbd-4534-a852-79ebbc81f947",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
